### ISSUE TITLE
Pulsar can't load the customized SerDe

### ISSUE THREAD

**Issue by sijie** [apache/pulsar#5350]
**Describe the bug**

Pulsar couldn't find deployed SerDe class. But it exists in my fat jar

```
12:59:00.903 [t1/n1/JournalBatchParseFunction-0] ERROR org.apache.pulsar.functions.instance.JavaInstanceRunnable - [t1/n1/JournalBatchParseFunction:0] Uncaught exception in Java Instance
java.lang.RuntimeException: User class must be in class path
        at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:67) ~[org.apache.pulsar-pulsar-functions-utils-2.4.1.jar:2.4.1]
        at org.apache.pulsar.functions.instance.InstanceUtils.createInstance(InstanceUtils.java:84) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:?]
        at org.apache.pulsar.functions.instance.InstanceUtils.initializeSerDe(InstanceUtils.java:43) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:?]
        at org.apache.pulsar.functions.source.TopicSchema.newSchemaInstance(TopicSchema.java:197) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:2.4.1]
        at org.apache.pulsar.functions.source.TopicSchema.lambda$getSchema$0(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:2.4.1]
        at java.util.HashMap.computeIfAbsent(HashMap.java:1127) ~[?:1.8.0_212]
        at org.apache.pulsar.functions.source.TopicSchema.getSchema(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:2.4.1]
        at org.apache.pulsar.functions.sink.PulsarSink.initializeSchema(PulsarSink.java:330) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:2.4.1]
        at org.apache.pulsar.functions.sink.PulsarSink.open(PulsarSink.java:255) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:2.4.1]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupOutput(JavaInstanceRunnable.java:769) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupJavaInstance(JavaInstanceRunnable.java:203) ~[org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:?]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:234) [org.apache.pulsar-pulsar-functions-instance-2.4.1.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
Caused by: java.lang.ClassNotFoundException: ru.alfabank.api.utils.JournalBatchSerde
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_212]
        at java.lang.Class.forName0(Native Method) ~[?:1.8.0_212]
        at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_212]
        at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:65) ~[org.apache.pulsar-pulsar-functions-utils-2.4.1.jar:2.4.1]
        ... 12 more
```

The command used to submit the function

```
bin/pulsar-admin --admin-url http://corpint4:6680 functions create --retain-ordering --jar http://binary/artifactory/libs-release-local/ru/alfabank/api/utils/pulsar-util-function-api-pack/0.10.2/pulsar-util-function-api-pack-0.10.2-all.jar --classname ru.alfabank.api.utils.JournalBatchParseFunction --inputs persistent://t1/n1/queue_1 --output persistent://t1/n1/queue_journal --output-serde-classname ru.alfabank.api.utils.JournalBatchSerde --tenant t1 --namespace n1
```

**Comment by sijie** [comment]
The problem is in PulsarSink.java. We don't use the right class loader to load the SerDe class.

https://github.com/apache/pulsar/blob/6ff1bbae0fe230eb72d62574d7685a745f884416/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java#L330

The fix should be straightforward to use `functionClassLoader` to load the SerDe class.

**Comment by Antti-Kaikkonen** [comment]
@sijie @wolfstudy 
I'm getting the same error when using --custom-serde-inputs:
```
./pulsar-admin functions create --jar test-functions/PulsarPOJOFunction-1.0-SNAPSHOT-jar-with-dependencies.jar --classname com.example.pulsartest.PojoFunction --name pojo-function --inputs user --output user2 --custom-serde-inputs "{"user":"com.example.pulsartest.UserSerDe"}"
```
Error found in logs:
```
09:48:40.471 [public/default/pojo-function-0] ERROR org.apache.pulsar.functions.instance.JavaInstanceRunnable - [public/default/pojo-function8:0] Uncaught exception in Java Instance
java.lang.RuntimeException: User class must be in class path
        at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:67) ~[org.apache.pulsar-pulsar-functions-utils-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.instance.InstanceUtils.createInstance(InstanceUtils.java:84) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.instance.InstanceUtils.initializeSerDe(InstanceUtils.java:43) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.source.TopicSchema.newSchemaInstance(TopicSchema.java:202) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.source.TopicSchema.newSchemaInstance(TopicSchema.java:210) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.source.TopicSchema.lambda$getSchema$0(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at java.util.HashMap.computeIfAbsent(HashMap.java:1133) ~[?:?]
        at org.apache.pulsar.functions.source.TopicSchema.getSchema(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.source.PulsarSource.lambda$setupConsumerConfigs$5(PulsarSource.java:163) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at java.util.TreeMap.forEach(TreeMap.java:1002) ~[?:?]
        at org.apache.pulsar.functions.source.PulsarSource.setupConsumerConfigs(PulsarSource.java:160) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.source.PulsarSource.open(PulsarSource.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupInput(JavaInstanceRunnable.java:732) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupJavaInstance(JavaInstanceRunnable.java:215) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:244) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
        at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: java.lang.ClassNotFoundException: com.example.pulsartest.UserSerDe
        at java.net.URLClassLoader.findClass(URLClassLoader.java:471) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:588) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:521) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:398) ~[?:?]
        at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:65) ~[org.apache.pulsar-pulsar-functions-utils-2.5.0.jar:2.5.0]
        ... 15 more
```
But com.example.pulsartest.UserSerDe exists in my jar. Pulsar is is 2.5.0 standalone.

**Comment by vlemaitre** [comment]
Hello,
I face the same issue on 2.5.0 version.
My Serde classes are into the jar file but the classes are not found when the function starts.
This issue is still present in 2.5.0.
Error logs :

14:52:24.290 [main] INFO  org.apache.pulsar.functions.LocalRunner - RuntimeSpawner quit because of
java.lang.RuntimeException: User class must be in class path
	at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:67) ~[org.apache.pulsar-pulsar-functions-utils-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.instance.InstanceUtils.createInstance(InstanceUtils.java:84) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.instance.InstanceUtils.initializeSerDe(InstanceUtils.java:43) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.source.TopicSchema.newSchemaInstance(TopicSchema.java:202) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.source.TopicSchema.newSchemaInstance(TopicSchema.java:210) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.source.TopicSchema.lambda$getSchema$0(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at java.util.HashMap.computeIfAbsent(HashMap.java:1127) ~[?:1.8.0_232]
	at org.apache.pulsar.functions.source.TopicSchema.getSchema(TopicSchema.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.source.PulsarSource.lambda$setupConsumerConfigs$5(PulsarSource.java:163) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at java.util.TreeMap.forEach(TreeMap.java:1005) ~[?:1.8.0_232]
	at org.apache.pulsar.functions.source.PulsarSource.setupConsumerConfigs(PulsarSource.java:160) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.source.PulsarSource.open(PulsarSource.java:66) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupInput(JavaInstanceRunnable.java:732) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.setupJavaInstance(JavaInstanceRunnable.java:215) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at org.apache.pulsar.functions.instance.JavaInstanceRunnable.run(JavaInstanceRunnable.java:244) ~[org.apache.pulsar-pulsar-functions-instance-2.5.0.jar:2.5.0]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_232]
Caused by: java.lang.ClassNotFoundException: com.hardisgroup.n1.poc.ExampleCreationSerde
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[?:1.8.0_232]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_232]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352) ~[?:1.8.0_232]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_232]
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_232]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_232]
	at org.apache.pulsar.functions.utils.Reflections.createInstance(Reflections.java:65) ~[org.apache.pulsar-pulsar-functions-utils-2.5.0.jar:2.5.0]
	... 15 more

**Comment by sijie** [comment]
reopen this issue to investigate the problem further.

**Comment by uatach** [comment]
I've seen this error, I mixed topics and ended up setting the serde input on a topic that had an avro schema.

**Comment by tisonkun** [comment]
Closed as stale. Please open a new issue if it's still relevant to maintain versions.

**Comment by JeffBolle** [comment]
I am also encountering this on Pulsar 2.11.0

**Comment by tisonkun** [comment]
@JeffBolle you may open a new issue with repro (better a demo repo).

**Comment by gmiklos-ltg** [comment]
There is also an issue with PulsarSource that relates to TopicSchema not using the correct classloader. I would like to suggest a fix in a PR that I will create shortly.

**Comment by tisonkun** [comment]
@gmiklos-ltg do you mean https://github.com/apache/pulsar/pull/20090? Thanks for your patch and I'll try to review it in days :)

**Comment by fishisnow** [comment]
I am also encountering this on Pulsar 3.0.0, how do i skip this error, it block my function instance create on my cluster

**Comment by tisonkun** [comment]
It's fixed on master only now - that means, 3.0.0 doesn't contain the fix.
